pyigm/clustering/xcorr_utils.py

```diff
@@ -222,3 +225,136 @@ def clean_matrix(W,n=-99):
     W = np.where(np.isnan(W),n,W)
     W = np.where(np.isinf(W),n,W)
     return W
+
+
+def random_gal(galreal, Nrand, Nmin=20, DZ=0.01, smooth_scale=10.):
+    """
+    Prefered random galaxy generator. For a given galaxy with a
+    given magnitude (and other properties), it calculates the redshift
+    sensitivity function from galaxies in a magnitude band around the
+    selected one (i.e., including slightly brighter and fainter
+    galaxies), and places Nrand new galaxies at a random redshift given
+    by a smoothed version of the observed sensitivity function. For
+    extremely bright or faint galaxies (rare) the sensitivity function
+    is calculated from at least Nmin (=20) galaxies (i.e. the magnitude
+    band is increased).
+
+    Parameters
+    ----------
+    galreal
+    Nrand
+    Nmin : int, optional
+    DZ : float, optional
+      delta z for the histogram for getting the
+    smooth_scale : float, optional
+       smoothing scale for the histogram (in number of bins, so depends on DZ)
+
+    Returns
+    -------
+    galrand : np rec array
+      Copy of galreal, Nrand times
+    """
+    from pyigm.clustering.randist import RanDist
+    from scipy.interpolate import CubicSpline
+    from scipy.ndimage import gaussian_filter as gf
+
+    debug = 0
+    Ckms = 299792.458
+    Nmin = int(Nmin)  # minimum number of galaxys for the fit
+    zmin = np.max([np.min(galreal.ZGAL[galreal.ZGAL > 0]), 1e-9])
+    zmax = np.max(galreal.ZGAL)
+    # spline in z
+    galreal.sort(order='MAG')  # np.recarray.sort()
+    galrand = galreal.repeat(Nrand)
+    delta_mag = 0.5
+
+    bins = np.append(np.linspace(0, zmin, 20), np.arange(zmin + DZ, zmax + 10 * DZ, DZ))
+    # bins = np.arange(0, zmax+DZ, DZ)
+
+    # Make subhistograms depending on magnitude. Use dictionary.
+    VALS = dict()
+    SPL = dict()
+    aux_hist, _ = np.histogram(galreal.ZGAL, bins)
+    VALS['all'] = gf(aux_hist.astype(float), smooth_scale)  # smooth the histogram
+    #SPL['all'] = InterpCubicSpline(0.5 * (bins[:-1] + bins[1:]), VALS['all'].astype(float))
+    SPL['all'] = CubicSpline(0.5 * (bins[:-1] + bins[1:]), VALS['all'].astype(float))
+
+    delta_mag2 = delta_mag
+    magbins = np.arange(17, 26, delta_mag2)
```

Comment:
limits somewhat arbitrary, should generalize (?)

Reply:
have allowed for flexible input on this now.  
And a magbins option too.

Reply:
Scratch that, linear step size in delta_mag is part of the
code for now.